### ISSUE TITLE
Script disconnected unexpectedly

### ISSUE THREAD

**Issue by wonderphil** [hashicorp/packer#5646]
Keep getting `Script Disconnected unexpectedly` when trying to build an Amazon EC2 ubuntu 16.04 instance.  packer script is [here](https://gist.github.com/wonderphil/e6ed20860a5b9cfa76102e3aa40faa74)

If i runn all the commands by hand against an instance it works fine no issue, but every time packer run we get the above error, with logging we can see:

```
packer: 2017/11/28 16:52:58 Remote command exited without exit status or exit signal.
2017/11/28 16:52:58 packer: 2017/11/28 16:52:58 [INFO] RPC endpoint: Communicator ended with: 2300218
2017/11/28 16:52:58 [INFO] 1382076 bytes written for 'stdout'
2017/11/28 16:52:58 [INFO] 2266 bytes written for 'stderr'
2017/11/28 16:52:58 [INFO] RPC client: Communicator ended with: 2300218
2017/11/28 16:52:58 [INFO] RPC endpoint: Communicator ended with: 2300218
2017/11/28 16:52:58 packer: 2017/11/28 16:52:58 [INFO] 1382076 bytes written for 'stdout'
2017/11/28 16:52:58 packer: 2017/11/28 16:52:58 [INFO] 2266 bytes written for 'stderr'
2017/11/28 16:52:58 packer: 2017/11/28 16:52:58 [INFO] RPC client: Communicator ended with: 2300218
```

- Packer version `Packer v1.1.0`
- Running packer on `Ubuntu 14.04.5 LTS` to build `Ubuntu 16.04 LTS`
- Debug log output from [here](https://gist.github.com/wonderphil/c1f8ce753c95d2ff99ab1938596dab17)
  
- The _simplest example template and scripts_ needed to reproduce the bug.
 [here](https://gist.github.com/wonderphil/e6ed20860a5b9cfa76102e3aa40faa74)

**Comment by rickard-von-essen** [comment]
You restart ssh in the middle of your script. If you split into two parts, ending the first just after the restart and mark that part with `expect_disconnect` I think it should work.

**Comment by wonderphil** [comment]
We dont do any restarts, the ansible step that it seems to fail on every time is installing Perl patch command below, which when run by hand takes less then a min and doesnt drop ssh or reload shell:  

```
/bin/bash -c 'umask 022; /usr/bin/env SHELL=/bin/bash PERLBREW_ROOT=/home/user/perl5/perlbrew /bin/perlbrew install-patchperl' user
```

That said we tried adding 

```
      "expect_disconnect": true,
      "start_retry_timeout": "30m",
```

to the provisioner and this time, we see it reconnect and end:

```
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 Remote command exited without exit status or exit signal.
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 [INFO] RPC endpoint: Communicator ended with: 2300218
2017/11/28 17:48:45 [INFO] 1285629 bytes written for 'stdout'
2017/11/28 17:48:45 [INFO] 1761 bytes written for 'stderr'
2017/11/28 17:48:45 [INFO] RPC client: Communicator ended with: 2300218
2017/11/28 17:48:45 [INFO] RPC endpoint: Communicator ended with: 2300218
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 [INFO] 1285629 bytes written for 'stdout'
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 [INFO] 1761 bytes written for 'stderr'
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 [INFO] RPC client: Communicator ended with: 2300218
2017/11/28 17:48:45 ui:     amazon:             "st
    amazon:             "st
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 opening new ssh session
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 ssh session open error: 'EOF', attempting reconnect
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 reconnecting to TCP connection for SSH
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 handshaking with SSH
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 handshake complete!
2017/11/28 17:48:45 packer: 2017/11/28 17:48:45 opening new ssh session
2017/11/28 17:48:46 packer: 2017/11/28 17:48:46 [INFO] agent forwarding enabled
2017/11/28 17:48:46 packer: 2017/11/28 17:48:46 starting remote command: rm -f /tmp/script_3169.sh
2017/11/28 17:48:46 packer: 2017/11/28 17:48:46 [INFO] RPC endpoint: Communicator ended with: 0
2017/11/28 17:48:46 [INFO] RPC client: Communicator ended with: 0
2017/11/28 17:48:46 [INFO] RPC endpoint: Communicator ended with: 0
2017/11/28 17:48:46 packer: 2017/11/28 17:48:46 [INFO] RPC client: Communicator ended with: 0
2017/11/28 17:48:46 [INFO] (telemetry) ending shell
```

**Comment by rickard-von-essen** [comment]
Line 90?

**Comment by wonderphil** [comment]
but its failing at line 115, it get pass line 90 no issues

**Comment by SwampDragons** [comment]
@wonderphil would you mind setting `expect_disconnect` anyway just as a sanity check?

**Comment by jamestwebber** [comment]
I've had this issue as well, in the past and currently. When I added `expect_disconnect` packer will exit "successfully", it just doesn't finish the installation script!

In the past I think the build worked with a larger instance..unfortunately now there is an issue with Grub demanding interaction if I switch to a c5.

**Comment by jamestwebber** [comment]
In my case I think this is just a matter of the builder spending a while downloading some package (via `conda install`) and packer decides it's had enough waiting and shuts it down. It would be great if I could just extend the timeout limit to something longer but I haven't been able to find such a thing.

**Comment by mwhooker** [comment]
If the remote end disconnects packer, then it's unlikely the rest of the script will be executed correctly. It's best to separate your provision scripts by restart boundaries. You can use `start_retry_timeout` and `pause_before` to tune this behavior.

There's an open ticket here about keepalives which may be relevant: https://github.com/hashicorp/packer/issues/4777

You may want to make sure your ssh server isn't configured to disconnect idle clients

**Comment by mwhooker** [comment]
@wonderphil Is this still an issue for you? It looks like the error has gone away once `expect_disconnect` was added.

**Comment by mwhooker** [comment]
@jamestwebber I'm not sure I follow. Why don't you create a new issue with all the required debugging info and we'll see what we can do.

**Comment by rickard-von-essen** [comment]
> On my end, I tried adding some ServerKeepAlive to my ssh_config but had the same result:

@jamestwebber that is a config file for OpenSSH client, Packer uses a native Go ssh library and don't read/understand any config except for the one in the packer template.

**Comment by jamestwebber** [comment]
That makes sense. And I will create a new issue for my image.

**Comment by mwhooker** [comment]
I believe this is solved in #5830, but I would appreciate your help testing if you're able to build that branch.
